### PR TITLE
Provide Amazon SAM integration for jaxrs based applications

### PR DESCRIPTION
**Pull request by evanchooly**
renamed the lambda-resteasy module to amazon-sam and refactored atop the vertx virtual connection bits

### CODE REVIEW

File: extensions/amazon-sam/runtime/src/main/java/io/quarkus/amazon/sam/SamHandler.java
Comment (stuartwdouglas) [inline extensions/amazon-sam/runtime/src/main/java/io/quarkus/amazon/sam/SamHandler.java]: Can you give a bit of explanation as to why the AWS proxy implementation is mixed in with the SAM implementation? As far as I can tell from the AWS docs they are really seperate concepts so I don't really understand why this is all mixed together?

**Comment by evanchooly** [comment]
That's not the gateway handler.  That's the bootstrap code to spin up the
quarkus app so that there is a process to feed requests to.  SAM doesn't
need the gateway proxy in this case.

File: extensions/amazon-sam/runtime/src/main/java/io/quarkus/amazon/sam/SamHandler.java
Comment (patriot1burke) [inline extensions/amazon-sam/runtime/src/main/java/io/quarkus/amazon/sam/SamHandler.java]: Again, you do not handle non-text types.  See stuarts implementation of this:

https://github.com/quarkusio/quarkus/blob/c58a1c1eace83126de787b5003f11a619227500c/extensions/amazon-lambda-http/runtime/src/main/java/io/quarkus/amazon/lambda/http/runtime/AwsHttpHandler.java#L82

File: extensions/amazon-sam/runtime/src/main/java/io/quarkus/amazon/sam/SamHandler.java
Comment (patriot1burke) [inline extensions/amazon-sam/runtime/src/main/java/io/quarkus/amazon/sam/SamHandler.java]: This does not handle non-text content.  See Stuarts implementation here:

https://github.com/quarkusio/quarkus/blob/c58a1c1eace83126de787b5003f11a619227500c/extensions/amazon-lambda-http/runtime/src/main/java/io/quarkus/amazon/lambda/http/runtime/AwsHttpHandler.java#L127

**Review by patriot1burke — CHANGES_REQUESTED** [review]

**Comment by patriot1burke** [comment]
I don't know how the amazon-lambda module is supposed to be used. I understand it exists so that Arc can instantiate the handler class.  Is it for Graal only lambda deployments?  How can a non-Graal deployment use this?

File: extensions/amazon-sam/deployment/pom.xml
Comment (evanchooly) [reply extensions/amazon-sam/deployment/pom.xml]: This isn't related to Amazon Lambda, though.  This is the SAM approach which is different than lambda

File: extensions/amazon-sam/deployment/src/main/java/io/quarkus/amazon/sam/deployment/AmazonSamProcessor.java
Comment (evanchooly) [reply extensions/amazon-sam/deployment/src/main/java/io/quarkus/amazon/sam/deployment/AmazonSamProcessor.java]: hrm.  curious.  the tests all pass so i'll see why *that* is and update accordingly.

**Comment by patriot1burke** [comment]
I don't know how the amazon-lambda module is supposed to be used. I understand it exists so that Arc can instantiate the handler class. Is it for Graal only lambda deployments? How can a non-Graal deployment use this?

**Comment by patriot1burke** [comment]
After all these issues are resolved, this will need documentation.

File: extensions/amazon-sam/runtime/src/main/java/io/quarkus/amazon/sam/SamHandler.java
Comment (evanchooly) [reply extensions/amazon-sam/runtime/src/main/java/io/quarkus/amazon/sam/SamHandler.java]: To be honest this code should be used in the sam module, too, but it doesn't exist on master.  what gives?

File: integration-tests/amazon-sam/pom.xml
Comment (patriot1burke) [inline integration-tests/amazon-sam/pom.xml]: This pom fails to compile when I checked out your branch.  It could not find this dependency.  Might be because i "skipped tests".

**Comment by evanchooly** [comment]
native mode was on my agenda today.  but, yeah, docs need to be written soon.

File: integration-tests/amazon-sam/pom.xml
Comment (evanchooly) [reply integration-tests/amazon-sam/pom.xml]: run `mvn install` from the devtools folder

File: devtools/maven/pom.xml
Comment (gastaldi) [inline devtools/maven/pom.xml]: Parent POM defines 3.1.0, is there any specific reason why 3.1.1 is needed here?

File: integration-tests/amazon-sam/pom.xml
Comment (gastaldi) [inline integration-tests/amazon-sam/pom.xml]:
```suggestion
        <dependency>
            <groupId>org.junit.jupiter</groupId>
            <artifactId>junit-jupiter</artifactId>
            <scope>test</scope>
        </dependency>
```

File: extensions/amazon-sam/deployment/pom.xml
Comment (patriot1burke) [inline extensions/amazon-sam/deployment/pom.xml]: This project should not depend on Resteasy.  it should be an additional dependency that is added to the application's pom.xml file after the HTTP api is chose (quarkus-resteasy, quarkus-vertx-web, or quarkus-undertow)

**Review by patriot1burke — CHANGES_REQUESTED** [review]

**Comment by stuartwdouglas** [comment]
Does this mean it is no longer possible to use the AWS Proxy gateway without using SAM? I don't really understand why these are being combined, as they seem to me to be independent concepts.

File: extensions/amazon-lambda-resteasy/runtime/src/test/java/io/quarkus/amazon/lambda/resteasy/adapter/test/ResteasyAwsProxyTest.java
Comment (stuartwdouglas) [inline extensions/amazon-lambda-resteasy/runtime/src/test/java/io/quarkus/amazon/lambda/resteasy/adapter/test/ResteasyAwsProxyTest.java]: These tests should not be deleted, in my draft I adapted them to the virtual channel. These tests picked up quite a few different potential problems that can be caused by using the API incorrectly.

File: extensions/amazon-sam/runtime/src/main/java/io/quarkus/amazon/sam/SamHandler.java
Comment (stuartwdouglas) [inline extensions/amazon-sam/runtime/src/main/java/io/quarkus/amazon/sam/SamHandler.java]: You need to use the multi value header map to make sure that multiple instances of headers are handled properly.

File: extensions/amazon-sam/runtime/src/main/java/io/quarkus/amazon/sam/SamHandler.java
Comment (stuartwdouglas) [inline extensions/amazon-sam/runtime/src/main/java/io/quarkus/amazon/sam/SamHandler.java]: If the request has a content-length header you can use that. Otherwise the size of the first bit of content (content.content().readableBytes()) is not a bad guess.

**Review by stuartwdouglas — DISMISSED** [review]

File: integration-tests/amazon-sam/pom.xml
Comment (stuartwdouglas) [inline integration-tests/amazon-sam/pom.xml]: These tests don't seem to run for me

File: integration-tests/amazon-sam/src/test/resources/projects/configure-existing-template/pom.xml
Comment (stuartwdouglas) [inline integration-tests/amazon-sam/src/test/resources/projects/configure-existing-template/pom.xml]: AFAIK need to be in src/it, and have an invoker.properties file to actually run the tests. See the maven plugin for an example.

Comment (stuartwdouglas) [reply integration-tests/amazon-sam/src/test/resources/projects/configure-existing-template/pom.xml]: Actually I am wrong, the invoker plugin should be removed as it is not needed, but the test needs to be renamed to GenerateSamConfigTestCase so maven will run it.

**Comment by geoand** [comment]
Can you please rebase onto the latest `master`? 

There is a CI fix that should take care of the error reported.

Thanks

**Comment by evanchooly** [comment]
> Can you please rebase onto the latest `master`?
> 
> There is a CI fix that should take care of the error reported.
> 
> Thanks

done.

**Comment by gsmet** [comment]
CI is failing so we won't merge this yet. BTW, could you please squash things into a single commit? Thanks.

**Comment by evanchooly** [comment]
I have some updates coming so it's not quite baked yet.  I tried to change the PR to draft but apparently can't.  I'll definitely squash before merging.

File: extensions/amazon-lambda-resteasy/runtime/src/test/java/io/quarkus/amazon/lambda/resteasy/adapter/test/ResteasyAwsProxyTest.java
Comment (evanchooly) [reply extensions/amazon-lambda-resteasy/runtime/src/test/java/io/quarkus/amazon/lambda/resteasy/adapter/test/ResteasyAwsProxyTest.java]: these tests as written on my branch don't work with the virtual channel bits.  we can copy your updated tests over, though.

File: extensions/amazon-sam/runtime/src/main/java/io/quarkus/amazon/sam/SamHandler.java
Comment (evanchooly) [reply extensions/amazon-sam/runtime/src/main/java/io/quarkus/amazon/sam/SamHandler.java]: it does the multivalue mapping internally.

**Comment by evanchooly** [comment]
Still working on that doc file so you can disregard its abrupt ending.  Can be committed separately as bill is needing *this* as a base for the native half.

**Review by patriot1burke — APPROVED** [review]

**Pull request merged**
